Makefile

```diff
@@ -91,3 +91,14 @@ debian-build-image:
 # Build Ubuntu build image
 ubuntu-build-image:
 	docker build -t kat-ubuntu-build-image packaging/ubuntu
+
+# Note about `pip-compile`: exported requirements.txt's target the Python version and platform used to run the command
```

Comment:
That's because pip currently does the dependency resolution, whereas with `pip-compile` the resulting `requirements.txt` file will already include all transitive dependencies. See https://github.com/jazzband/pip-tools/issues/1326 and https://github.com/pypa/pip/issues/9644 for the limitations of this.

Reply:
You still don't explain how these limitations make it very difficult for us to maintain. With regards to the last issue we should run pip install with `--no-deps`, given that pip-compile already resolved the dependencies it's not necessary that pip does that again anyway.

With regards to the limitations, if we use python 3.11 to generate the requirements (and also have a check in CI for that in case somebody would use a different Python version) and install with Python 3.11 in the containers, what would make it very difficult to maintain? For the Debian packages that use older Python versions there could be an issue, but that is still theoretical for now. In my opinion we should use Debian packages for Python dependencies, but even if we continue with the current packaged virtualenvs we could easily fix any issue in the Debian packaging scripts. We already give different options to pip install there anyway because of the mismatch between installing wheels from PyPI and what Debian packaging expects.

Reply:
As far as I understand, the `requirements.txt` with resolved dependencies that someone generates on an Apple Silicon Mac (Darwin AArch64) can be different from the resolved dependencies on a x86-64 Linux machine. This means that we'd have to keep track of a separate `requirements.txt` per pair of platform and architecture.

For other, older Python versions we could indeed re-run `pip-compile` when needed, but when we don't pin exact dependency versions in `pyproject.toml` we will probably end up with slightly different versions which is also not desirable.

Reply:
It can be in theory, but with our dependencies, that doesn't seem to be the case if you look at what poetry generates for keiko/octopoes. I think it is unlikely that we will ever run into this given that what we do doesn't have any specific dependencies on architecture or platform.

Reply:
We discussed this in the office last Thursday. While the architecture is unlikely to be a problem for OpenKAT, that pip-compile does not take Python versions into account will likely be an issue. An example is pytest only depending on the backported exceptiongroup package for Python 3.10 and earlier. There might be workarounds for this, but the whole reason for going with pip-compile was that it was simpler and having to workaround these kind issues would not make it so simple anymore. So it would better to use Poetry which does support dependencies for only specific Python versions.